mistletoe/markdown_renderer.py

```diff
@@ -0,0 +1,537 @@
+        return obj
+
+    def __init__(self, matches):
+        self.children = list(map(LinkReferenceDefinition, matches))
+
+
+class Fragment:
+    """
+    Markdown fragment. Used when rendering trees of span tokens into flat sequences.
+    May carry additional data in addition to the text.
+
+    Attributes:
+        text (str): markdown fragment
+    """
+
+    def __init__(self, text: str, **extras):
+        self.text = text
+        for k, v in extras.items():
+            setattr(self, k, v)
+
+
+class MarkdownRenderer(BaseRenderer):
+    """
+    Markdown renderer.
+
+    Designed to make as "clean" a roundtrip as possible, markdown -> parsing -> rendering -> markdown,
+    except for nonessential whitespace. Except when rendering with word wrapping enabled.
+
+    Includes HTMLBlocks and HTMLSpans in the parsing.
+    """
+
+    _whitespace = re.compile(r"\s+")
+
+    def __init__(self, *extras):
+        block_token.remove_token(block_token.Footnote)
+        super().__init__(
+            *chain(
+                (
+                    block_token.HTMLBlock,
+                    span_token.HTMLSpan,
+                    BlankLine,
+                    LinkReferenceDefinitionBlock,
+                ),
+                extras,
+            )
+        )
+        self.render_map["SetextHeading"] = self.render_setext_heading
+        self.render_map["CodeFence"] = self.render_fenced_code_block
+        self.render_map[
+            "LinkReferenceDefinition"
+        ] = self.render_link_reference_definition
+
+    def render(self, token: token.Token, max_line_length: int = None) -> str:
+        """
+        Renders the tree of tokens rooted at the given token into markdown.
+
+        If `max_line_length` is specified, the document is word wrapped to the
+        specified line length. Otherwise the formatting from the original (parsed)
+        document is retained as much as possible.
+        """
+        if isinstance(token, block_token.BlockToken):
+            lines = self.render_map[token.__class__.__name__](
+                token, max_line_length=max_line_length
+            )
+        else:
+            lines = self.span_to_lines([token], max_line_length=max_line_length)
+
+        return "".join(map(lambda line: line + "\n", lines))
+
+    # rendering of span/inline tokens.
+    # rendered into sequences of Fragments.
+
+    def render_raw_text(self, token) -> Iterable[Fragment]:
+        yield Fragment(token.content, wordwrap=True)
+
+    def render_strong(self, token: span_token.Strong) -> Iterable[Fragment]:
+        return self.embed_span(Fragment(token.delimiter * 2), token.children)
+
+    def render_emphasis(self, token: span_token.Emphasis) -> Iterable[Fragment]:
+        return self.embed_span(Fragment(token.delimiter), token.children)
+
+    def render_inline_code(self, token: span_token.InlineCode) -> Iterable[Fragment]:
+        yield Fragment(token.delimiter + token.raw_content + token.delimiter)
```

Comment:
Using `InlineCode.raw_content` here, while it makes sense, is a bit problematic.  Because we need to think about this when altering the AST like in the nice example you prepared in the `dev-guide.md` - i.e. changing inner `RawText` instance has no effect in the example when used with the `MarkdownRenderer`, so results may be surprising.

I think the main complication here is that we decided, I believe correctly, to do some preprocessing of the inline code according to the spec globally, namely this line within `InlineCode.__init__()`:

```python
content = content.replace('\n', ' ')
```

Any ideas of how to get out of this "conflicting stuff" elegantly?

Reply:
I agree that it's a problem -- it would be much better to use the content of the RawText node instead of this new and rather hidden field. Only, we can't do that if the RawText content is processed.

The preprocessing in `InlineCode.__init__()` consists of two steps: first we convert line breaks to space, and then we strip one leading and one trailing space if the string begins AND ends with spaces and is not a spaces-only string. This makes the Markdown to HTML rendering spec-compliant, so of course the processing is needed in that process.

One option is to move the processing into `HTMLRenderer.render_inline_code()`. The other renderers (LaTeX, Jira, XWiki20, AST, Scheme) will also need to be updated accordingly -- I don't know what the expected behavior is for them, but that should be possible to find out (and it's easy to just keep the current behavior). It will be an API break but not a severe one.

Elegant or not, this is the least bad solution I can think of :)

Reply:
Good points, thank you. Or, I have got an idea now that we could temporarily disable the preprocessing of inline code - within the `with MarkdownRenderer() as renderer` block (by overriding the `__enter__` and `__exit__` methods, plus introducing a special boolean flag variable). That could do the trick? :)

Reply:
Solved the issue by allowing inline code to be re-flowed, that is, not preserving the line breaks. While this is against the principle of preserving the formatting as far as possible, it is more in line with how span tokens work in general, and with how inline code is handled when rendering to HTML.

An option could be to allow the InlineToken to hold a sequence of RawText and LineBreak tokens, instead of a single RawText node. (Possible future improvement?)